Changes to support Item Kits, support for MySQL 5.7, and a tweak for …

SteveIreland opened the pull request:
This is a much improved version of the earlier MySQL 5.7 pull request.  Hopefully you'll agree.  If not please let me know what I've missed.

odiea commented:
Mystery Stock Report still shows.
Item models and Sales models sum need to be changed to max.
Register still shows unable to add item when using a kit.
When does the kit discount come into effect?
Except for the above issues it appears to work fine in MySQL 5.7 and php 7.

jekkos commented:
Hi Steve, thanks for recreating the PR, looks better now.
Can you check indentation as well? I don't know which IDE you use (PHPStorm?) but in this case you can easily configure to use tabs instead of spaces.

SteveIreland commented:
@odiea Thanks for the feedback.  I'm working on these issues now.

@jekkos I've changed PhpStorm back to use tabs instead of spaces, but I thought I read something on this site that told me I was supposed to be using spaces. I can't find it now but it's possible I somehow navigated to some other repository and read some guidelines that aren't used here.  I think I need to reformat the source files to convert them back from space to tabs, so I'll do that too.

jekkos commented:
@SteveIreland You are right the information on the wiki wasn't correct. initially we said spaces but to be in line with CI3 we chose tabs after all. If you check the diff's on github then you'll see that there's some trouble there. I adapted the wiki to reflect that.

jekkos commented:
I think it just refers to CI3 guidelines now (for the indentation part)

SteveIreland commented:
I've updated the pull request.

The mystery stock report option should be resolved now.

The error with adding a kit item that doesn't have a definition on the item table has been fixed.

Discounts are now being applied as follows.  A discount can be applied at the system level.  A customer discount will override the the system discount.  If there is a kit discount and it is greater than the customer discount then it will be applied, otherwise the customer discount will be applied.

Space indentation has been converted to tab indentation.  (I hope.)

@odiea My tests are showing that the sum function is the correct choice, so my test scenario isn't covering the test scenario you are working with.  Can  you elaborate on what test case scenario isn't working correctly?   Thanks.

There are still a couple of changes I need to make and will continue on those tomorrow.

odiea commented:
Stock report is gone.
These are the issues I see in Items and Takings.

This is what I see when I load Items. Look at the QTY column.
![2017-01-22 04_43_56-photos](https://cloud.githubusercontent.com/assets/10962177/22182255/09ee0464-e05e-11e6-907b-289a59a07dd2.png)
This is what I see when I load Items in earlier version. Look at the QTY column.
![2017-01-22 04_46_04-](https://cloud.githubusercontent.com/assets/10962177/22182262/46e4e00e-e05e-11e6-960b-d7e553c3f732.png)
Your Takings.
![2017-01-22 04_45_16-wedidit _ powered by ospos 3 0 2](https://cloud.githubusercontent.com/assets/10962177/22182265/54c72c22-e05e-11e6-8610-7d1fdff8fa56.png)
Earlier Takings.
![2017-01-22 04_47_04-wedidit _ powered by ospos 3 0 2](https://cloud.githubusercontent.com/assets/10962177/22182268/60be09e2-e05e-11e6-945d-e05464f7e769.png)

SteveIreland commented:
@odiea The pull request has been updated with a fix for those two issues you identified.

SteveIreland commented:
@odiea @daN4cat  I've made the remaining changes from SUM to MAX, renamed the database upgrade file, and corrected the two php files that were still using space tabbing in some spots of the code instead of tabs.

jekkos commented:
So @odiea you tested this does it look alright? I might want to have a look as well this weekend hopefully I can find some time.

SteveIreland commented:
I have some more changes coming to it soon.  I'll do my best to try to have them posted Saturday.  I'm trying to slip in the feature to have the option to not include item kit items on the invoice/receipt but to still have the item  kit itself included.

daN4cat commented:
Let us know when you have completed the work so we can pull the PR and commit to master for a community test time.
I'll see if I find the time too to come back to this project and close the parts I need to deploy vs 3 to my wife's shop system.

jekkos commented:
@daN4cat good idea if there are no known bugs left we can merge indeed and await community feedback.

odiea commented:
Other than the learning curve the only other minor issue was in Detailed Reports after you do an update the table doesn't update until after you do a refresh.

jekkos commented:
I have deployed an instance of this PR atg http://web.opensourcepos.5bc0857a.svc.dockerapp.io/home

free and easy to test now.. I'll give it a try right away

odiea commented:
I assume this is Php 7 with MySQL 5.7?

jekkos commented:
Good point, I'll redeploy the stack with PHP7 and new mysql

odiea commented:
I believe the current version is 100% Php 7 compatible. It will be the future commits that will create the issues.

SteveIreland commented:
I've committed my last set of changes. (I was a bit tired and failed to change the description of the commit, but it was more than to "Correct a typo" - sorry about that.)

Notable changes are that I reduced the number of invoice and receipt line sort sequence from 4 to 3 (Entry, Group by Type, and Group by Category).  During these last changes I became a bit more familiar with the software and found that Entry sequence is actually the current sort sequence and Entry will be the default.

The new Print option is now supported for receipts and invoices is sorted.    You can flag it for all kit items to be listed, or only those that have a zero price, or only those that are stocked items (i.e. materials for service work).  That was an educational experience for me.  If you look at the code and think it could be coded a bit more elegantly then please let me know ... especially if you can direct me to an existing example of how it should be done.

daN4cat commented on application/models/Item.php:
```diff
@@ -60,13 +60,58 @@ public function get_found_rows($search, $filters)
-		$this->db->from('items');
-		$this->db->join('suppliers', 'suppliers.person_id = items.supplier_id', 'left');
-		$this->db->join('inventory', 'inventory.trans_items = items.item_id');
+		$this->db->select('max(items.name) as name');
```
This could have probably been one select statement as we use to do in other models so you avoid having multiple functions calls. Also SQL functions are in general in capital letters to make a distinction between functions and variables.

daN4cat commented on application/models/reports/Detailed_receivings.php:
```diff
@@ -39,22 +39,43 @@ public function getDataColumns()
 	public function getDataByReceivingId($receiving_id)
 	{
-		$this->db->select('receiving_id, receiving_date, SUM(quantity_purchased) AS items_purchased, CONCAT(employee.first_name, " ", employee.last_name) AS employee_name, supplier.company_name AS supplier_name, SUM(subtotal) AS subtotal, SUM(total) AS total, SUM(profit) AS profit, payment_type, comment, reference');
+		$this->db->select('receiving_id, 
```
good idea, thanks!

daN4cat commented on application/models/reports/Specific_customer.php:
```diff
@@ -44,29 +44,29 @@ public function getDataColumns()
 	public function getData(array $inputs)
 	{
-		$this->db->select('sale_id, sale_date, SUM(quantity_purchased) AS items_purchased, employee_name, SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit, payment_type, comment');
+		$this->db->select('sale_id, MAX(sale_date) AS sale_date, SUM(quantity_purchased) AS items_purchased, MAX(employee_name) AS employee_name, SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit, MAX(payment_type) AS payment_type, MAX(comment) AS comment');
```
it would be nice to explode this statement in multiple lines as you did in other cases.

daN4cat commented on application/models/reports/Specific_discount.php:
```diff
@@ -40,12 +40,12 @@ public function getDataColumns()
 	public function getData(array $inputs)
 	{
-		$this->db->select('sale_id, sale_date, SUM(quantity_purchased) AS items_purchased, customer_name, SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit, payment_type, comment');
+		$this->db->select('sale_id, MAX(sale_date) AS sales_date, SUM(quantity_purchased) AS items_purchased, MAX(customer_name) AS customer_name, SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit, MAX(payment_type) AS payment_type, MAX(comment) AS comment');
```
same here

daN4cat commented on database/database.sql:
```diff
@@ -4,9 +4,9 @@
 --
```
As said already once, these changes should go to tables.sql because if we run the grunt script all of this will be lost as database.sql is regenerated.
You can leave this as it is, but please make sure tables.sql contains the same change.

daN4cat replied:
In case you want to install the development tools, see https://github.com/jekkos/opensourcepos/wiki/Development-setup as I updated the instructions what to install before you run the bower & grunt scripts as per next points in that wiki.

SteveIreland replied:
Okay, thanks.

daN4cat commented:
@SteveIreland you probably need to rebase your code with the latest changes.
Somehow merging a change a lot of duplicated history went in, but you should have not many conflicts.

SteveIreland commented:
Okay .... remind me never to commit to an epic change again.  ;-)

daN4cat commented:
Oh well, you went for three in one.... sooo :-)

daN4cat commented:
By the way, after your set of changes we will need to commit to master and stabilise for a while as most likely there will be bit and pieces not right here and there with all the latest merged changes.

daN4cat commented:
@SteveIreland just rebase again the two bug fixes I just submitted. I stop here so you can work with no further changes.
When done I will take and merge your changes to master and then we all test and fix issues from there.

SteveIreland commented:
Working on a proper rebase now ... will continue in a few hours.

daN4cat commented:
Take it easy, I know I made your Sunday busier than expected... sorry.

SteveIreland commented:
Okay, I've gone through the cycle and then noticed a message from you that I missed.  SO I'll try another rebase after fetching the master changes again.  If it goes flawlessly  I'll have it done.  However, if not then I will probably not be able to finish until tomorrow.  I'll let you know.

SteveIreland commented:
Okay after reviewing what I did, I think I might have unintentionally accomplished what you wanted.  I rebased against a fresh resync of master and I think I picked up your changes.  I had to -force the update to the item-kit branch, so that might have had some unintentional consequences.

daN4cat commented:
I'll check this evening and merge in master the code so we avoid any more merging/rebasing stories. From that we can fix any bug issue we found and stabilise as I suggested earlier.

jekkos commented:
Force push after rebase might indeed have been necessary as you were  probably rewriting some of the git history (replaying your commits on top of the latest master one). At least if it's in a PR this is of no harm and advantage is that history will look cleaner. 
I wanted to work on the dotenv PR but will wait until this one is integrated.

odiea commented:
I downloaded and installed your branch and I see nothing and with inspecting elements I just get error 500 when trying to load the login page.

daN4cat commented:
@jekkos before the dotenv there is the picture PR.

jekkos commented:
@daN4cat Ok but that one  needs more work if I'm not mistaken? the dotenv shouldn't impact other branches that much anyway.

jekkos commented:
Ok no this PR is more up to date. might be better to merge it first indeeed.

odiea commented:
Let me know when I can test this. It was not working from the download on his site.

daN4cat commented:
I'm merging and committing to master. If broken we all fix it :-)

SteveIreland commented:
I'm not sure if anyone here would recognize a reference to a TV "Bar Rescue" episode, but ....

"I'm all in!"

daN4cat commented:
@SteveIreland it's now in master but you need to rebase your branch because I fixed models/sale.php as containing an uncompleted merge code. I'm not sure what I did was the correct thing, so you need to verify.

Rest of the group, all to the testing seats now....

odiea commented:
Should there be 2 Line Sequence here?

![2017-01-30 13_23_09-](https://cloud.githubusercontent.com/assets/10962177/22440467/dfd7c34c-e6f0-11e6-8912-32f698553364.png)

odiea commented:
I also see you found the following in Models sale.
![2017-01-30 13_36_25-_c__users_pavilion_downloads_opensourcepos-item-kit_kit6_application_models_sale](https://cloud.githubusercontent.com/assets/10962177/22440542/32a08082-e6f1-11e6-991b-09e6f4bd7faa.png)

SteveIreland commented:
There should only be one Line Sequence config option.

SteveIreland commented:
I'll start fixing what's reported in about 2 hours.  

Is the proper way to do that to start a new branch for fixes?

daN4cat commented:
Maybe starting a new branch off latest master will be cleaner. You can use the old for comparison purposes to check what was the code before the latest merge.
